lib/Subscriber.js

```diff
@@ -0,0 +1,592 @@
+    {
+      logger.warn('Event header does not match SUBSCRIBE');
+      request.reply(489);
+      this._dialogTerminated(C.RECEIVE_BAD_NOTIFY);
+
+      return;
+    }
+
+    // Process Subscription-State header.
+    const subs_state = request.parseHeader('subscription-state');
+
+    if (!subs_state) 
+    {
+      logger.warn('missed Subscription-State header');
+      request.reply(400);
+      this._dialogTerminated(C.RECEIVE_BAD_NOTIFY);
+
+      return;
+    }
+
+    request.reply(200);
+
+    const new_state = this._stateStringToNumber(subs_state.state);
+    const prev_state = this._state;
+ 
+    if (prev_state !== C.STATE_TERMINATED && new_state !== C.STATE_TERMINATED) 
+    {
+      this._state = new_state;
+
+      if (subs_state.expires !== undefined) 
+      {
+        const expires = subs_state.expires;
+        const expires_timestamp = new Date().getTime() + (expires * 1000);
+        const max_time_deviation = 2000;
+
+        // Expiration time is shorter and the difference is not too small.
+        if (this._expires_timestamp - expires_timestamp > max_time_deviation) 
+        {
+          logger.debug('update sending re-SUBSCRIBE time');
+
+          this._scheduleSubscribe(expires);
+        }
+      }
+    }
+
+    if (prev_state !== C.STATE_PENDING && new_state === C.STATE_PENDING) 
+    {
+      logger.debug('emit "pending"');
+      this.emit('pending');
+    }
+    else if (prev_state !== C.STATE_ACTIVE && new_state === C.STATE_ACTIVE) 
+    {
+      logger.debug('emit "active"');
+      this.emit('active');
+    }
+
+    const body = request.body;
+
+    // Check if the notify is final.
+    const is_final = new_state === C.STATE_TERMINATED;
+
+    // Notify event fired only for notify with body.
+    if (body) 
+    {
+      const content_type = request.getHeader('content-type');
+
+      logger.debug('emit "notify"');
+      this.emit('notify', is_final, request, body, content_type);
+    }
+
+    if (is_final)
+    {
+      const reason = subs_state.reason;
+      let retry_after = undefined;
+
+      if (subs_state.params && subs_state.params['retry-after'] !== undefined)
+      {
+        retry_after = parseInt(subs_state.params['retry-after']);
+      } 
+
+      this._dialogTerminated(C.RECEIVE_FINAL_NOTIFY, reason, retry_after);
+    }
+  }
+
+  /**
+   * User API
+   */
+
+  /** 
+   * Send the initial (non-fetch)  and subsequent subscribe.
+   * @param {string} body - subscribe request body.
+   */
+  subscribe(body = null) 
+  {
+    logger.debug('subscribe()');
+
+    if (this._state === C.STATE_INIT)
+    {
+      this._sendInitialSubscribe(body, this._headers);
+    }
+    else 
+    {
+      this._sendSubsequentSubscribe(body, this._headers);
+    }
+  }
+
+  /** 
+   * terminate. 
+   * Send un-subscribe or fetch-subscribe (with Expires: 0).
+   * @param {string} body - un-subscribe request body
+   */
+  terminate(body = null) 
+  {
+    logger.debug('terminate()');
+
```

Comment:
Avoid calling this method multiple times. This should be the last public method to be called in the instance.

```suggestion
if (this._terminated)
{
  return;
}

this.terminated = true;
```

Reply:
Better than the above would be checking:

`this._state === C.STATE_TERMINATED`

Reply:
It could be better, but ... not according RFC 6665.

subscriber switched to terminated state not after sending un-subscribe, but 
after receive terminated notify or after receive non ok response to subscribe.

See diagram: [https://datatracker.ietf.org/doc/html/rfc6665](url)
page 10.

So I used your "above" way:
To prevent duplicate terminate() call I used property **this._send_unsubscribe**
(instead this._terminated).
If you think that it's more clear I can change the name to **_terminated**.